django/db/models/fields/json.py

```diff
@@ -0,0 +1,489 @@
+    }
+    _default_hint = ('dict', '{}')
+
+    def __init__(self, encoder=None, decoder=None, default=dict, *args, **kwargs):
+        self.encoder, self.decoder = encoder, decoder
+        super().__init__(default=default, *args, **kwargs)
+
+    def check(self, **kwargs):
+        return [
+            *super().check(**kwargs),
+            *self._check_json_support(),
+        ]
+
+    @classmethod
+    def _check_json_support(cls):
+        errors = []
+        if not default_connection.features.supports_json:
+            errors.append(
+                checks.Error(
+                    _('JSONField is not supported by this database backend.'),
+                    obj=cls
+                )
+            )
+        return errors
+
+    def deconstruct(self):
+        name, path, args, kwargs = super().deconstruct()
+        if self.default is dict:
+            del kwargs['default']
+        if self.encoder is not None:
+            kwargs['encoder'] = self.encoder
+        if self.decoder is not None:
+            kwargs['decoder'] = self.decoder
+        return name, path, args, kwargs
+
+    def from_db_value(self, value, expression, connection):
+        if value is None:
+            return value
+        elif connection.features.interprets_empty_strings_as_nulls and value == '':
+            return None
+        else:
+            try:
+                return json.loads(value, cls=self.decoder)
+            except json.decoder.JSONDecodeError:
+                return value
+
+    def get_internal_type(self):
+        return 'JSONField'
+
+    def get_prep_value(self, value):
+        if value is None:
+            return value
+        return json.dumps(value, cls=self.encoder)
+
+    def get_transform(self, name):
+        transform = super().get_transform(name)
+        if transform:
+            return transform
+        return KeyTransformFactory(name)
+
+    def select_format(self, compiler, sql, params):
+        if compiler.connection.vendor == 'postgresql':
+            # Avoid psycopg2's automatic decoding to allow custom decoder
+            return '%s::text' % sql, params
+        return super().select_format(compiler, sql, params)
+
+    def validate(self, value, model_instance):
+        super().validate(value, model_instance)
+        try:
+            json.dumps(value, cls=self.encoder)
+        except TypeError:
+            raise exceptions.ValidationError(
+                self.error_messages['invalid'],
+                code='invalid',
+                params={'value': value},
+            )
+
+    def value_to_string(self, obj):
+        return self.value_from_object(obj)
+
+    def formfield(self, **kwargs):
+        return super().formfield(**{
+            'form_class': forms.JSONField,
+            'encoder': self.encoder,
+            'decoder': self.decoder,
+            **kwargs,
+        })
+
+
+class SimpleFunctionOperatorMixin(FieldGetDbPrepValueMixin):
```

Comment:
It feels like `django.db.models.lookups` should be made into a package (`base`, `datetime`, `text`) and this should live in `.json`.

Reply:
I think so, too, but that sounds like it should be in another PR, doesn't it?

Reply:
Yes. It might be easier to do after this lands, but someone else may decide otherwise.